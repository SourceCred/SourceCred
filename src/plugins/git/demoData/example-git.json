{
    "commits": {
        "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f": {
            "hash": "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
            "treeHash": "7be3ecfee5314ffa9b2d93fc4377792b2d6d70ed"
        },
        "69c5aad50eec8f2a0a07c988c3b283a6490eb45b": {
            "hash": "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
            "treeHash": "bbf3b8b3d26a4f884b5c022d46851f593d329192"
        },
        "8d287c3bfbf8455ef30187bf5153ffc1b6eef268": {
            "hash": "8d287c3bfbf8455ef30187bf5153ffc1b6eef268",
            "treeHash": "3dfb84795e07341b05fad3a0d5a55f8304b2d7d8"
        },
        "c08ee3a4edea384d5291ffcbf06724a13ed72325": {
            "hash": "c08ee3a4edea384d5291ffcbf06724a13ed72325",
            "treeHash": "2f7155e359fd0ecb96ffdca66fa45b6ed5792809"
        },
        "c2b51945e7457546912a8ce158ed9d294558d294": {
            "hash": "c2b51945e7457546912a8ce158ed9d294558d294",
            "treeHash": "bdff5d94193170015d6cbb549b7b630649428b1f"
        },
        "d160cca97611e9dfed642522ad44408d0292e8ea": {
            "hash": "d160cca97611e9dfed642522ad44408d0292e8ea",
            "treeHash": "569e1d383759903134df75230d63c0090196d4cb"
        },
        "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc": {
            "hash": "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc",
            "treeHash": "819fc546cea489476ce8dc90785e9ba7753d0a8f"
        }
    },
    "trees": {
        "2f7155e359fd0ecb96ffdca66fa45b6ed5792809": {
            "entries": {
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                }
            },
            "hash": "2f7155e359fd0ecb96ffdca66fa45b6ed5792809"
        },
        "3dfb84795e07341b05fad3a0d5a55f8304b2d7d8": {
            "entries": {
                ".gitmodules": {
                    "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
                    "name": ".gitmodules",
                    "type": "blob"
                },
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "pygravitydefier": {
                    "hash": "762c062fbdc7ec198cd693e95d55b374a08ff3e3",
                    "name": "pygravitydefier",
                    "type": "commit"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                }
            },
            "hash": "3dfb84795e07341b05fad3a0d5a55f8304b2d7d8"
        },
        "569e1d383759903134df75230d63c0090196d4cb": {
            "entries": {
                ".gitmodules": {
                    "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
                    "name": ".gitmodules",
                    "type": "blob"
                },
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "TODOS.txt": {
                    "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
                    "name": "TODOS.txt",
                    "type": "blob"
                },
                "pygravitydefier": {
                    "hash": "762c062fbdc7ec198cd693e95d55b374a08ff3e3",
                    "name": "pygravitydefier",
                    "type": "commit"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                },
                "src": {
                    "hash": "7b79d579b62994faba3b69fdf8aa442586c32681",
                    "name": "src",
                    "type": "tree"
                }
            },
            "hash": "569e1d383759903134df75230d63c0090196d4cb"
        },
        "78fc9c83023386854c6bfdc5761c0e58f68e226f": {
            "entries": {
                "index.py": {
                    "hash": "674b0b476989384510304846248b3acd16206782",
                    "name": "index.py",
                    "type": "blob"
                },
                "quantum_gravity.py": {
                    "hash": "aea4f28abb23abde151b0ead4063227f8bf6c0b0",
                    "name": "quantum_gravity.py",
                    "type": "blob"
                }
            },
            "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f"
        },
        "7b79d579b62994faba3b69fdf8aa442586c32681": {
            "entries": {
                "index.py": {
                    "hash": "674b0b476989384510304846248b3acd16206782",
                    "name": "index.py",
                    "type": "blob"
                },
                "quantum_gravity.py": {
                    "hash": "887ad856bbc1373da146106c86cb581ad78cdafe",
                    "name": "quantum_gravity.py",
                    "type": "blob"
                }
            },
            "hash": "7b79d579b62994faba3b69fdf8aa442586c32681"
        },
        "7be3ecfee5314ffa9b2d93fc4377792b2d6d70ed": {
            "entries": {
                ".gitmodules": {
                    "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
                    "name": ".gitmodules",
                    "type": "blob"
                },
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "pygravitydefier": {
                    "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
                    "name": "pygravitydefier",
                    "type": "commit"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                },
                "src": {
                    "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f",
                    "name": "src",
                    "type": "tree"
                }
            },
            "hash": "7be3ecfee5314ffa9b2d93fc4377792b2d6d70ed"
        },
        "819fc546cea489476ce8dc90785e9ba7753d0a8f": {
            "entries": {
                ".gitmodules": {
                    "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
                    "name": ".gitmodules",
                    "type": "blob"
                },
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "TODOS.txt": {
                    "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
                    "name": "TODOS.txt",
                    "type": "blob"
                },
                "pygravitydefier": {
                    "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
                    "name": "pygravitydefier",
                    "type": "commit"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                },
                "src": {
                    "hash": "7b79d579b62994faba3b69fdf8aa442586c32681",
                    "name": "src",
                    "type": "tree"
                }
            },
            "hash": "819fc546cea489476ce8dc90785e9ba7753d0a8f"
        },
        "bbf3b8b3d26a4f884b5c022d46851f593d329192": {
            "entries": {
                ".gitmodules": {
                    "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
                    "name": ".gitmodules",
                    "type": "blob"
                },
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                },
                "TODOS.txt": {
                    "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
                    "name": "TODOS.txt",
                    "type": "blob"
                },
                "pygravitydefier": {
                    "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
                    "name": "pygravitydefier",
                    "type": "commit"
                },
                "science.txt": {
                    "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
                    "name": "science.txt",
                    "type": "blob"
                },
                "src": {
                    "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f",
                    "name": "src",
                    "type": "tree"
                }
            },
            "hash": "bbf3b8b3d26a4f884b5c022d46851f593d329192"
        },
        "bdff5d94193170015d6cbb549b7b630649428b1f": {
            "entries": {
                "README.txt": {
                    "hash": "0fb31858c8e3710be77e1dbb8880acf8a5543d82",
                    "name": "README.txt",
                    "type": "blob"
                }
            },
            "hash": "bdff5d94193170015d6cbb549b7b630649428b1f"
        }
    }
}
