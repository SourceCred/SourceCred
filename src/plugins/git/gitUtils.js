// @flow

import {execFileSync} from "child_process";
import fs from "fs";
import mkdirp from "mkdirp";
import path from "path";

export interface Utils {
  exec: GitDriver;
  head(): string;
  writeAndStage(filename: string, contents: string): void;
  deterministicCommit(message: string): void;
}

export type GitDriver = (args: string[], options?: ExecOptions) => string;
// `ExecOptions` is the type of the second argument to `execFileSync`.
// See here for details: https://nodejs.org/api/child_process.html
type ExecOptions = Object;

export function localGit(repositoryPath: string): GitDriver {
  return function git(args: string[], options?: ExecOptions): string {
    return execFileSync(
      "git",
      ["-C", repositoryPath, ...args],
      options
    ).toString();
  };
}

export function makeUtils(repositoryPath: string): Utils {
  const git = localGit(repositoryPath);
  return {
    exec: git,

    head() {
      return git(["rev-parse", "HEAD"]).trim();
    },

    writeAndStage(filename: string, contents: string) {
      const filepath = path.join(repositoryPath, filename);
      const dirpath = path.join(repositoryPath, path.dirname(filename));
      mkdirp.sync(dirpath);
      fs.writeFileSync(filepath, contents);
      git(["add", filename]);
    },

    deterministicCommit(message: string): void {
      git(
        [
          "-c",
          "user.name=Test Runner",
          "-c",
          "user.email=nobody@example.com",
          "commit",
          "-m",
          message,
        ],
        {
          env: {
            TZ: "UTC",
            GIT_AUTHOR_DATE: "2001-02-03T04:05:06",
            GIT_COMMITTER_DATE: "2002-03-04T05:06:07",
          },
        }
      );
    },
  };
}
