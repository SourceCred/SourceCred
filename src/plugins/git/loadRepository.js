/*
 * Load a git repository into memory. This dumps the commit and tree
 * data into a structured form. Contents of blobs are not loaded.
 *
 * If the repository contains file names that are not valid UTF-8
 * strings, the result is undefined.
 *
 * Note: git(1) is a runtime dependency of this module.
 */
// @flow

import {execFileSync} from "child_process";

import type {Repository, Hash, Commit, Tree, TreeEntry} from "./types";

export type GitDriver = (args: string[], options?: ExecOptions) => string;
type ExecOptions = Object; // close enough
export function localGit(repositoryPath: string): GitDriver {
  return function git(args: string[], options?: ExecOptions): string {
    // Throws an Error on shell failure.
    return execFileSync(
      "git",
      ["-C", repositoryPath, ...args],
      options
    ).toString();
  };
}

/**
 * Load a Git repository from disk into memory. The `rootRef` should be
 * a revision reference as accepted by `git rev-parse`: "HEAD" and
 * "origin/master" will be common, while a specific SHA or tag might be
 * used to fix a particular state of a repository.
 */
export function loadRepository(
  repositoryPath: string,
  rootRef: string
): Repository {
  const git = localGit(repositoryPath);
  const commits = findCommits(git, rootRef);
  const trees = findTrees(git, new Set(commits.map((x) => x.treeHash)));
  return {commits: hashMap(commits), trees: hashMap(trees)};
}

function hashMap<T: {+hash: Hash}>(ts: $ReadOnlyArray<T>): Map<Hash, T> {
  const result = new Map();
  ts.forEach((t) => {
    result.set(t.hash, t);
  });
  return result;
}

function findCommits(git: GitDriver, rootRef: string): Commit[] {
  return git(["log", "--oneline", "--pretty=%H %T", rootRef])
    .split("\n")
    .filter((line) => line.length > 0)
    .map((line) => {
      const [hash, treeHash] = line.split(" ");
      return {hash, treeHash};
    });
}

function findTrees(git: GitDriver, rootTrees: Set<Hash>): Tree[] {
  const result: Tree[] = [];
  const visited: Set<Hash> = new Set();
  const frontier: Set<Hash> = new Set(rootTrees);
  while (frontier.size > 0) {
    const next = frontier.values().next();
    if (next.done) {
      // Flow doesn't know that this is impossible, but it is.
      throw new Error("Impossible! `frontier` had positive size.");
    }
    const treeHash: Hash = next.value;
    visited.add(treeHash);
    frontier.delete(treeHash);
    const tree = loadTree(git, treeHash);
    result.push(tree);
    for (const entry of tree.entries.values()) {
      if (entry.type === "tree" && !visited.has(entry.hash)) {
        frontier.add(entry.hash);
      }
    }
  }
  return result;
}

function loadTree(git: GitDriver, treeHash: Hash): Tree {
  const entries: TreeEntry[] = git(["ls-tree", "--full-tree", "-z", treeHash])
    .split("\0")
    .filter((line) => line.length > 0)
    .map((line) => {
      // See `git help ls-tree`, section OUTPUT FORMAT, for details.
      const [metadata, name] = line.split("\t");
      const [mode, type, hash] = metadata.split(" ");
      if (type !== "blob" && type !== "commit" && type !== "tree") {
        throw new Error(
          `entry ${treeHash}[${JSON.stringify(name)}] ` +
            `has unexpected type "${type}"`
        );
      }
      return {name, type, hash};
    });
  return {hash: treeHash, entries: new Map(entries.map((e) => [e.name, e]))};
}
