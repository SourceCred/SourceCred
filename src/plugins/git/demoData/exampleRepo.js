// @flow

import mkdirp from "mkdirp";

import {makeUtils} from "../gitUtils";
import type {Hash} from "../types";

type RepositoryInfo = {|
  +path: string,
  +commits: $ReadOnlyArray<Hash>, // in oldest-to-newest order
|};

export function createExampleRepo(intoDirectory: string): RepositoryInfo {
  const repositoryPath = intoDirectory;
  mkdirp(repositoryPath);
  const git = makeUtils(repositoryPath);
  const commits = [];

  git.exec(["init"]);

  git.writeAndStage("README.txt", "Amazing physics going on...\n");
  git.deterministicCommit("Initial commit");
  commits.push(git.head());

  git.writeAndStage("src/index.py", "import antigravity\n");
  git.writeAndStage(
    "src/quantum_gravity.py",
    'raise NotImplementedError("TODO(physicists)")\n'
  );
  git.writeAndStage("TODOS.txt", "1. Resolve quantum gravity\n");
  git.deterministicCommit("Discover gravity");
  commits.push(git.head());

  git.writeAndStage(
    "src/quantum_gravity.py",
    "import random\nif random.random() < 0.5:\n  import antigravity\n"
  );
  git.deterministicCommit("Solve quantum gravity");
  commits.push(git.head());

  git.exec(["rm", "TODOS.txt"]);
  git.deterministicCommit("Clean up TODOS");
  commits.push(git.head());

  return {path: repositoryPath, commits};
}
