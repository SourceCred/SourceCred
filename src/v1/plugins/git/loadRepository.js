/*
 * Load a git repository into memory. This dumps the commit and tree
 * data into a structured form. Contents of blobs are not loaded.
 *
 * If the repository contains file names that are not valid UTF-8
 * strings, the result is undefined.
 *
 * Note: git(1) is a runtime dependency of this module.
 */
// @flow

import type {GitDriver} from "./gitUtils";
import type {Repository, Hash, Commit, Tree, TreeEntry} from "./types";
import {localGit} from "./gitUtils";

/**
 * Load a Git repository from disk into memory. The `rootRef` should be
 * a revision reference as accepted by `git rev-parse`: "HEAD" and
 * "origin/master" will be common, while a specific SHA or tag might be
 * used to fix a particular state of a repository.
 */
export function loadRepository(
  repositoryPath: string,
  rootRef: string
): Repository {
  const git = localGit(repositoryPath);
  const commits = findCommits(git, rootRef);
  const trees = findTrees(git, new Set(commits.map((x) => x.treeHash)));
  return {commits: objectMap(commits), trees: objectMap(trees)};
}

function objectMap<T: {+hash: Hash}>(ts: $ReadOnlyArray<T>): {[Hash]: T} {
  const result = {};
  ts.forEach((t) => {
    result[t.hash] = t;
  });
  return result;
}

function findCommits(git: GitDriver, rootRef: string): Commit[] {
  return git(["log", "--oneline", "--pretty=%H %T %P", rootRef])
    .split("\n")
    .filter((line) => line.length > 0)
    .map((line) => {
      const [hash, treeHash, ...parentHashes] = line.trim().split(" ");
      const submoduleUrls = loadSubmoduleUrls(git, hash);
      return {hash, parentHashes, treeHash, submoduleUrls};
    });
}

const GITMODULES_SUBMODULES_KEY_RE = /^submodule\.(.*)\.(path|url)$/;

function loadSubmoduleUrls(
  git: GitDriver,
  commitHash: Hash
): {[path: string]: string} {
  const gitmodulesRef = `${commitHash}:.gitmodules`;
  const gitmodulesBlob: string | null = (() => {
    try {
      return git(["rev-parse", "--quiet", "--verify", gitmodulesRef]).trim();
    } catch (e) {
      if (e.status === 1) {
        // No .gitmodules file here.
        return null;
      } else {
        throw e;
      }
    }
  })();
  if (gitmodulesBlob == null) {
    // No problem; there just weren't any submodules at this commit.
    return {};
  }

  // The output format of the following is `${key}\n${value}\0...`, as
  // specified in `git help config`'s section about the `-z` option.
  // The format is safe because keys are strictly validated; see the
  // function `git_config_parse_key` in `git/git:config.c`.
  const rawConfig = git(["config", "--blob", gitmodulesBlob, "--list", "-z"]);
  const configKeyValuePairs = rawConfig
    .split("\0")
    .filter((line) => line.length > 0)
    .map((line) => {
      const separator = line.indexOf("\n");
      if (separator < 0) {
        // Shouldn't happen, according to Git docs. Guard anyway.
        throw new Error(`Bad .gitmodules line at ${commitHash}: ${line}`);
      }
      return {
        key: line.substring(0, separator),
        value: line.substring(separator + 1),
      };
    });

  const submoduleInfoByKey: {
    [submoduleKey: string]: {path: string | null, url: string | null},
  } = {};
  configKeyValuePairs.forEach(({key, value}) => {
    const match = key.match(GITMODULES_SUBMODULES_KEY_RE);
    if (!match) {
      return;
    }
    const [_, submoduleKey, kind] = match;
    if (submoduleInfoByKey[submoduleKey] == null) {
      submoduleInfoByKey[submoduleKey] = {path: null, url: null};
    }
    if (kind !== "path" && kind !== "url") {
      throw new Error(`Invariant violation: bad kind: ${kind}`);
    }
    submoduleInfoByKey[submoduleKey][kind] = value;
  });

  const result = {};
  Object.keys(submoduleInfoByKey).forEach((submoduleKey) => {
    const {path, url} = submoduleInfoByKey[submoduleKey];
    if (path != null && url != null) {
      result[path] = url;
    } else {
      console.warn(`Partial submodule at ${commitHash}: ${submoduleKey}`);
    }
  });
  return result;
}

function findTrees(git: GitDriver, rootTrees: Set<Hash>): Tree[] {
  const result: Tree[] = [];
  const visited: Set<Hash> = new Set();
  const frontier: Set<Hash> = new Set(rootTrees);
  while (frontier.size > 0) {
    const next = frontier.values().next();
    if (next.done) {
      // Flow doesn't know that this is impossible, but it is.
      throw new Error("Impossible! `frontier` had positive size.");
    }
    const treeHash: Hash = next.value;
    visited.add(treeHash);
    frontier.delete(treeHash);
    const tree = loadTree(git, treeHash);
    result.push(tree);
    Object.keys(tree.entries).forEach((key) => {
      const entry = tree.entries[key];
      if (entry.type === "tree" && !visited.has(entry.hash)) {
        frontier.add(entry.hash);
      }
    });
  }
  return result;
}

function loadTree(git: GitDriver, treeHash: Hash): Tree {
  const entries: {[name: string]: TreeEntry} = {};
  git(["ls-tree", "--full-tree", "-z", treeHash])
    .split("\0")
    .filter((line) => line.length > 0)
    .forEach((line) => {
      // See `git help ls-tree`, section OUTPUT FORMAT, for details.
      const [metadata, name] = line.split("\t");
      const [_unused_mode, type, hash] = metadata.split(" ");
      if (type !== "blob" && type !== "commit" && type !== "tree") {
        throw new Error(
          `entry ${treeHash}[${JSON.stringify(name)}] ` +
            `has unexpected type "${type}"`
        );
      }
      entries[name] = {name, type, hash};
    });
  return {hash: treeHash, entries: entries};
}
