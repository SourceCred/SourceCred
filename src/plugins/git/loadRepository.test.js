// @flow

import tmp from "tmp";

import {createExampleRepo} from "./demoData/exampleRepo";
import {makeUtils} from "./gitUtils";
import {loadRepository} from "./loadRepository";

const cleanups: (() => void)[] = [];
afterAll(() => {
  cleanups.forEach((f) => {
    f();
  });
});

function mkdtemp() {
  const result = tmp.dirSync({unsafeCleanup: true});
  cleanups.push(() => result.removeCallback());
  return result.name;
}

describe("loadRepository", () => {
  it("loads from HEAD", () => {
    const repository = createExampleRepo(mkdtemp());
    expect(loadRepository(repository.path, "HEAD")).toMatchSnapshot();
  });

  it("processes an old commit", () => {
    const repository = createExampleRepo(mkdtemp());
    const whole = loadRepository(repository.path, "HEAD");
    const part = loadRepository(repository.path, repository.commits[1]);

    // Check that `part` is a subset of `whole`...
    for (const hash of part.commits.keys()) {
      expect(part.commits.get(hash)).toEqual(whole.commits.get(hash));
    }
    for (const hash of part.trees.keys()) {
      expect(part.trees.get(hash)).toEqual(whole.trees.get(hash));
    }

    // ...and that it's the right subset.
    expect({
      commits: new Set(part.commits.keys()),
      trees: new Set(part.trees.keys()),
    }).toMatchSnapshot();
  });
});
