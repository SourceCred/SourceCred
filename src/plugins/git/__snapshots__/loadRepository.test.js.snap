// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadRepository loads from HEAD 1`] = `
Object {
  "commits": Map {
    "3507b7c8690e329477c76b73f8511f78fdbafe52" => Object {
      "hash": "3507b7c8690e329477c76b73f8511f78fdbafe52",
      "treeHash": "aeb32ba2769e46068cc09aa05723b5dd81c2e3a0",
    },
    "1468ff8437299dc4aa426996784be30a87226007" => Object {
      "hash": "1468ff8437299dc4aa426996784be30a87226007",
      "treeHash": "77e66e35964b9c0a2cde52fb2640d3004b742feb",
    },
    "d6693032379c8c55aafc72ad355db29ce71396b0" => Object {
      "hash": "d6693032379c8c55aafc72ad355db29ce71396b0",
      "treeHash": "c32b1ddc3a622e7a7e30c0b3f56a78eaed19cfab",
    },
    "fb67eef11f7f34e4ad530ce36624c47fef0cd2e2" => Object {
      "hash": "fb67eef11f7f34e4ad530ce36624c47fef0cd2e2",
      "treeHash": "b936e6133d3316250f0a91779f72c72141279941",
    },
    "849572148969b2c3480e8479604ccba3b189092a" => Object {
      "hash": "849572148969b2c3480e8479604ccba3b189092a",
      "treeHash": "e5e3b135d4f7d03e9906aead28c7757aa2b676bd",
    },
    "db7608c2b4eba9ca72c6e3d636a7795cf9bcb97d" => Object {
      "hash": "db7608c2b4eba9ca72c6e3d636a7795cf9bcb97d",
      "treeHash": "5822dfcc81bf3ef8f99f68a21a33d07386fbd7b8",
    },
    "486bb61573b09069dc9a55023b06e72049f8556d" => Object {
      "hash": "486bb61573b09069dc9a55023b06e72049f8556d",
      "treeHash": "79ee2359849733d6a51fd1b5da1cebdbaa29254f",
    },
  },
  "trees": Map {
    "aeb32ba2769e46068cc09aa05723b5dd81c2e3a0" => Object {
      "entries": Map {
        ".gitmodules" => Object {
          "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
          "name": ".gitmodules",
          "type": "blob",
        },
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "pygravitydefier" => Object {
          "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
          "name": "pygravitydefier",
          "type": "commit",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
        "src" => Object {
          "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f",
          "name": "src",
          "type": "tree",
        },
      },
      "hash": "aeb32ba2769e46068cc09aa05723b5dd81c2e3a0",
    },
    "77e66e35964b9c0a2cde52fb2640d3004b742feb" => Object {
      "entries": Map {
        ".gitmodules" => Object {
          "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
          "name": ".gitmodules",
          "type": "blob",
        },
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "TODOS.txt" => Object {
          "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
          "name": "TODOS.txt",
          "type": "blob",
        },
        "pygravitydefier" => Object {
          "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
          "name": "pygravitydefier",
          "type": "commit",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
        "src" => Object {
          "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f",
          "name": "src",
          "type": "tree",
        },
      },
      "hash": "77e66e35964b9c0a2cde52fb2640d3004b742feb",
    },
    "c32b1ddc3a622e7a7e30c0b3f56a78eaed19cfab" => Object {
      "entries": Map {
        ".gitmodules" => Object {
          "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
          "name": ".gitmodules",
          "type": "blob",
        },
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "TODOS.txt" => Object {
          "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
          "name": "TODOS.txt",
          "type": "blob",
        },
        "pygravitydefier" => Object {
          "hash": "29ef158bc982733e2ba429fcf73e2f7562244188",
          "name": "pygravitydefier",
          "type": "commit",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
        "src" => Object {
          "hash": "7b79d579b62994faba3b69fdf8aa442586c32681",
          "name": "src",
          "type": "tree",
        },
      },
      "hash": "c32b1ddc3a622e7a7e30c0b3f56a78eaed19cfab",
    },
    "b936e6133d3316250f0a91779f72c72141279941" => Object {
      "entries": Map {
        ".gitmodules" => Object {
          "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
          "name": ".gitmodules",
          "type": "blob",
        },
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "TODOS.txt" => Object {
          "hash": "ddec7477206c30c31b81482e56b877a0b3c2638b",
          "name": "TODOS.txt",
          "type": "blob",
        },
        "pygravitydefier" => Object {
          "hash": "762c062fbdc7ec198cd693e95d55b374a08ff3e3",
          "name": "pygravitydefier",
          "type": "commit",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
        "src" => Object {
          "hash": "7b79d579b62994faba3b69fdf8aa442586c32681",
          "name": "src",
          "type": "tree",
        },
      },
      "hash": "b936e6133d3316250f0a91779f72c72141279941",
    },
    "e5e3b135d4f7d03e9906aead28c7757aa2b676bd" => Object {
      "entries": Map {
        ".gitmodules" => Object {
          "hash": "8c6cac301e763aa6d36466e0a775eb804be2c311",
          "name": ".gitmodules",
          "type": "blob",
        },
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "pygravitydefier" => Object {
          "hash": "762c062fbdc7ec198cd693e95d55b374a08ff3e3",
          "name": "pygravitydefier",
          "type": "commit",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
      },
      "hash": "e5e3b135d4f7d03e9906aead28c7757aa2b676bd",
    },
    "5822dfcc81bf3ef8f99f68a21a33d07386fbd7b8" => Object {
      "entries": Map {
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
        "science.txt" => Object {
          "hash": "f1f2514ca6d7a6a1a0511957021b1995bf9ace1c",
          "name": "science.txt",
          "type": "blob",
        },
      },
      "hash": "5822dfcc81bf3ef8f99f68a21a33d07386fbd7b8",
    },
    "79ee2359849733d6a51fd1b5da1cebdbaa29254f" => Object {
      "entries": Map {
        "README.txt" => Object {
          "hash": "6e70bfb555b270a017c743df4cd7d35d5337e0ef",
          "name": "README.txt",
          "type": "blob",
        },
      },
      "hash": "79ee2359849733d6a51fd1b5da1cebdbaa29254f",
    },
    "78fc9c83023386854c6bfdc5761c0e58f68e226f" => Object {
      "entries": Map {
        "index.py" => Object {
          "hash": "674b0b476989384510304846248b3acd16206782",
          "name": "index.py",
          "type": "blob",
        },
        "quantum_gravity.py" => Object {
          "hash": "aea4f28abb23abde151b0ead4063227f8bf6c0b0",
          "name": "quantum_gravity.py",
          "type": "blob",
        },
      },
      "hash": "78fc9c83023386854c6bfdc5761c0e58f68e226f",
    },
    "7b79d579b62994faba3b69fdf8aa442586c32681" => Object {
      "entries": Map {
        "index.py" => Object {
          "hash": "674b0b476989384510304846248b3acd16206782",
          "name": "index.py",
          "type": "blob",
        },
        "quantum_gravity.py" => Object {
          "hash": "887ad856bbc1373da146106c86cb581ad78cdafe",
          "name": "quantum_gravity.py",
          "type": "blob",
        },
      },
      "hash": "7b79d579b62994faba3b69fdf8aa442586c32681",
    },
  },
}
`;

exports[`loadRepository processes an old commit 1`] = `
Object {
  "commits": Set {
    "db7608c2b4eba9ca72c6e3d636a7795cf9bcb97d",
    "486bb61573b09069dc9a55023b06e72049f8556d",
  },
  "trees": Set {
    "5822dfcc81bf3ef8f99f68a21a33d07386fbd7b8",
    "79ee2359849733d6a51fd1b5da1cebdbaa29254f",
  },
}
`;
